File: BOJ/16236/A.py
import sys
from collections import deque
sys.stdin = open("D:\Projects\AA\BOJ\\16236\input.txt")
input = sys.stdin.readline

N = int(input().rstrip())

arr = []
dir = [(-1, 0), (0, -1), (0, 1), (1, 0)]
check = []

for i in range(N):
    line = list(map(int, input().rstrip().split()))
    arr.append(line)

baby = []
size = 2
eaten = 0
for i in range(N):
    for j in range(N):
        if arr[i][j] == 9:
            baby = [i, j]
            arr[i][j] = 0
time = 0


def BFS(x1, y1, d1):
    t = 0
    q = deque([(x1, y1, d1)])
    no_fish = True
    x2 = 0
    y2 = 0
    while q:
        current = q.popleft()
        x = current[0]
        y = current[1]
        d = current[2]
        break_flag = False
        for i in range(len(dir)):
            a = dir[i][0]
            b = dir[i][1]
            if (x+a >= N or x+a < 0) or (y+b >= N or y+b < 0) or check[x+a][y+b] == 1 or arr[x+a][y+b] > size:
                continue
            if arr[x+a][y+b] < size and arr[x+a][y+b] > 0:
                x2 = x+a
                y2 = y+b
                no_fish = False
                break_flag = True
                t = d+1
                break
            q.append((x+a, y+b, d+1))
            check[x+a][y+b] = 1
        if break_flag:
            break
        t += 1
    if no_fish:
        t = 0
    return (x2, y2, t)


def reset_check():
    global check
    check = [[0 for __ in range(N)] for _ in range(N)]


target = -1
while True:
    reset_check()
    temp = []
    result = BFS(baby[0], baby[1], 0)
    if result[2] == 0:
        break
    temp.append(result)
    reset_check()
    check[result[0]][result[1]] = 1
    another = BFS(baby[0], baby[1], 0)

    while(result[2] == another[2]):
        temp.append(another)
        reset_check()
        for i in range(len(temp)):
            check[temp[i][0]][temp[i][1]] = 1
        another = BFS(baby[0], baby[1], 0)

    temp.sort()

    result = temp[0]
    x = result[0]
    y = result[1]
    target = result[2]
    arr[x][y] = 0

    baby[0] = x
    baby[1] = y
    eaten += 1

    if eaten == size:
        eaten = 0
        size = size + 1
    time += target

print(time)

File: BOJ/16236/A.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class point
{
public:
    int x;
    int y;
    int dis;
    point(int a, int b, int c)
    {
        x = a;
        y = b;
        dis = c;
    }
};

vector<pair<int, int>> dir = {{-1, 0}, {0, -1}, {0, 1}, {1, 0}};
vector<vector<int>> arr;
vector<vector<int>> check;
vector<point> nxt;
vector<int> temp;
queue<point> *q;

bool mysort(point a, point b)
{
    if (a.x < b.x)
    {
        return true;
    }
    else if (a.x == b.x)
    {
        if (a.y < b.y)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    else
    {
        return false;
    }
}

int main()
{
    freopen("input.txt", "rt", stdin);
    cin.tie(0);
    ios::sync_with_stdio(0);
    int N;
    cin >> N;
    temp = vector<int>(N, 0);
    arr = vector<vector<int>>(N, temp);
    check = vector<vector<int>>(N, temp);
    int x, y;
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            int n;
            cin >> n;
            arr[i][j] = n;
            if (n == 9)
            {
                x = i;
                y = j;
            }
        }
    }
    bool help = false;
    int eaten = 0;
    int size = 2;
    int time = 0;
    arr[x][y] = 0;
    while (!help)
    {
        q = new queue<point>;
        point initial = point(x, y, 0);
        q->push(initial);
        check = vector<vector<int>>(N, temp);
        check[x][y] = 1;
        nxt = vector<point>();
        int closest = -1;
        while (q->size())
        {
            point current = q->front();
            q->pop();
            // check if there is a fish on current point available to eat
            if (arr[current.x][current.y] > 0 && arr[current.x][current.y] < size)
            {
                if (closest == -1)
                {
                    closest = current.dis;
                }
                else if (closest != -1 && current.dis > closest)
                {
                    break;
                }
                nxt.push_back(current);
                // arr[current.x][current.y] = 0;
                // eaten++;
                // // increase baby shark size
                // if (eaten == size)
                // {
                //     size++;

                //     eaten = 0;
                // }
            }
            for (int i = 0; i < 4; i++)
            {
                int a = dir[i].first;
                int b = dir[i].second;
                if (current.x + a >= 0 && current.x + a < N && current.y + b >= 0 && current.y + b < N && !check[current.x + a][current.y + b] && arr[current.x + a][current.y + b] <= size)
                {
                    check[current.x + a][current.y + b] = 1;
                    point temp_point = point(current.x + a, current.y + b, current.dis + 1);
                    q->push(temp_point);
                }
            }
        }
        if (!nxt.size())
        {
            help = true;
        }
        else
        {
            sort(nxt.begin(), nxt.end(), mysort);
            point eat = nxt[0];
            arr[eat.x][eat.y] = 0;
            time += eat.dis;
            eaten++;
            // increase baby shark size
            if (eaten == size)
            {
                size++;
                eaten = 0;
            }
            x = eat.x;
            y = eat.y;
        }
        free(q);
    }
    cout << time;
}